pkg/operator/ceph/object/bucket/s3-handlers.go

```diff
@@ -0,0 +1,243 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bucket
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/yard-turkey/lib-bucket-provisioner/pkg/provisioner/api/errors"
+	"k8s.io/apimachinery/pkg/util/json"
+)
+
+type S3Agent struct {
+	client *s3.S3
+}
+
+func NewS3Agent(accessKey, secretKey, endpoint string) (*S3Agent, error) {
+	const cephRegion = "us-east-1"
+
+	sess, err := session.NewSession(
+		aws.NewConfig().
+			WithRegion(cephRegion).
+			WithCredentials(credentials.NewStaticCredentials(accessKey, secretKey, "")).
+			WithEndpoint(endpoint).
+			WithS3ForcePathStyle(true).
+			WithMaxRetries(20).
+			WithDisableSSL(true))
+	if err != nil {
+		return nil, err
+	}
+	svc := s3.New(sess)
+	return &S3Agent{
+		client: svc,
+	}, nil
+}
+
+func (s S3Agent) CreateBucket(bktName string) error {
+	logger.Infof("creating bucket %q", bktName)
+	bucketInput := &s3.CreateBucketInput{
+		Bucket: &bktName,
+	}
+	_, err := s.client.CreateBucket(bucketInput)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			logger.Infof("DEBUG: after s3 call, ok=%v, aerr=%v", ok, aerr)
+			switch aerr.Code() {
+			case s3.ErrCodeBucketAlreadyExists:
+				msg := fmt.Sprintf("Bucket %q already exists", bktName)
+				logger.Errorf(msg)
+				return errors.NewBucketExistsError(msg)
+			case s3.ErrCodeBucketAlreadyOwnedByYou:
+				msg := fmt.Sprintf("Bucket %q already owned by you", bktName)
+				logger.Errorf(msg)
+				return errors.NewBucketExistsError(msg)
+			}
+		}
+		return fmt.Errorf("Bucket %q could not be created: %v", bktName, err)
+	}
+
+	logger.Infof("successfully created bucket %q", bktName)
+	return nil
+}
+
+////////////////
+// Policy
+////////////////
+
+type action string
+
+const (
```

Comment:
Is this copied from another repo? If so does it make any sense to import that repo?

Reply:
It looks like these come from here: http://docs.ceph.com/docs/master/radosgw/bucketpolicy/

Reply:
Interesting, it's surprising that we would need to define these.

Reply:
> it looks like these come from here ...

That's correct.  With the exception of `All`, these are all the Ceph Object supported S3 actions.  Rather than define an arbitrary subset, I thought it better to include them all should the package see more use later on.  A slice of the actions we're concerned with is defined for bucket provisioning.

Reply:
@copejon @travisn what do you think about moving these constants to somewhere in _pkg/operator/ceph/object/_?

Reply:
My initial thought is no because the names are fairly s3 specific. Maybe in a bucket lib issue we can try to generalize these concepts, particularly if we want to add some form of policy to an OBC.

Reply:
What's supported in this package is 1:1 with Ceph's S3 implementation, per the docs.

> Would these make sense to move into the bucket provisioning library? Or are these ceph-specific for what actions and policies are created?

The library is agnostic to the object store interface, including policy, so these wouldn't have any meaning there.